### PR TITLE
Use 'is not null' in some System.Windows.Forms files

### PR DESCRIPTION
**Pull request by gpetrou**
Contributes to https://github.com/dotnet/winforms/issues/3459

## Proposed changes

- Use 'is not null' in some System.Windows.Forms files.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/4747)

### CODE REVIEW

**Comment by AraHaan** [comment]
Cant you change all ``!=`` to ``is not``?

File: src/System.Windows.Forms/src/System/Windows/Forms/LinkLabel.LinkCollection.cs [hunk 1]
```diff
@@ -143,7 +143,7 @@ public Link Add(int start, int length, object linkData)
 
             public int Add(Link value)
             {
-                if (value != null && value.Length != 0)
+                if (value is not null && value.Length != 0)
```
Comment (AraHaan) [inline src/System.Windows.Forms/src/System/Windows/Forms/LinkLabel.LinkCollection.cs:146]: cant ``value.Length != 0`` be changed to ``value.Length is not 0``?

File: src/System.Windows.Forms/src/System/Windows/Forms/LinkLabel.cs [hunk 2]
```diff
@@ -720,7 +720,7 @@ private void EnsureRun(Graphics g)
                         //
 
                         Region visualRegion = new Region(visualRectangle);
-                        if (_links != null && _links.Count == 1)
+                        if (_links is not null && _links.Count == 1)
```
Comment (AraHaan) [inline src/System.Windows.Forms/src/System/Windows/Forms/LinkLabel.cs:723]: Cant ``_links.Count == 1`` be changed to ``_links.Count is 1``?

File: src/System.Windows.Forms/src/System/Windows/Forms/LinkLabel.cs [hunk 3]
```diff
@@ -869,12 +869,12 @@ private void InvalidateLink(Link link)
...
         private void InvalidateLinkFonts()
         {
-            if (_linkFont != null)
+            if (_linkFont is not null)
             {
                 _linkFont.Dispose();
             }
 
-            if (_hoverLinkFont != null && _hoverLinkFont != _linkFont)
+            if (_hoverLinkFont is not null && _hoverLinkFont != _linkFont)
```
Comment (AraHaan) [inline src/System.Windows.Forms/src/System/Windows/Forms/LinkLabel.cs:877]: ``_hoverLinkFont is not _linkFont``?

File: src/System.Windows.Forms/src/System/Windows/Forms/LinkLabel.cs [hunk 4]
```diff
@@ -1077,7 +1077,7 @@ protected override void OnMouseUp(MouseEventArgs e)
 
                     Link clicked = PointInLink(e.X, e.Y);
 
-                    if (clicked != null && clicked == FocusLink && clicked.Enabled)
+                    if (clicked is not null && clicked == FocusLink && clicked.Enabled)
```
Comment (AraHaan) [inline src/System.Windows.Forms/src/System/Windows/Forms/LinkLabel.cs:1080]: ``clicked is FocusLink``?

File: src/System.Windows.Forms/src/System/Windows/Forms/LinkUtilities.cs [hunk 5]
```diff
@@ -138,17 +138,17 @@ public static LinkBehavior GetIELinkBehavior()
...
             }
 
-            if (key != null)
+            if (key is not null)
             {
                 string s = (string)key.GetValue("Anchor Underline");
                 key.Close();
 
-                if (s != null && string.Compare(s, "no", true, CultureInfo.InvariantCulture) == 0)
+                if (s is not null && string.Compare(s, "no", true, CultureInfo.InvariantCulture) == 0)
```
Comment (AraHaan) [inline src/System.Windows.Forms/src/System/Windows/Forms/LinkUtilities.cs:146]: ``is 0`` at the end?

File: src/System.Windows.Forms/src/System/Windows/Forms/LinkUtilities.cs [hunk 6]
```diff
@@ -138,17 +138,17 @@ public static LinkBehavior GetIELinkBehavior()
...
                 {
                     return LinkBehavior.NeverUnderline;
                 }
 
-                if (s != null && string.Compare(s, "hover", true, CultureInfo.InvariantCulture) == 0)
+                if (s is not null && string.Compare(s, "hover", true, CultureInfo.InvariantCulture) == 0)
```
Comment (AraHaan) [inline src/System.Windows.Forms/src/System/Windows/Forms/LinkUtilities.cs:151]: ``is 0`` at the end?

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs [hunk 7]
```diff
@@ -1108,7 +1108,7 @@ public override string Text
         {
             get
             {
-                if (SelectionMode != SelectionMode.None && SelectedItem != null)
+                if (SelectionMode != SelectionMode.None && SelectedItem is not null)
```
Comment (AraHaan) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs:1111]: did you forget an ``is not`` here?

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs [hunk 8]
```diff
@@ -1131,7 +1131,7 @@ public override string Text
                 // Scan through the list items looking for the supplied text string.  If we find it,
                 // select it.
                 //
-                if (SelectionMode != SelectionMode.None && value != null && (SelectedItem is null || !value.Equals(GetItemText(SelectedItem))))
+                if (SelectionMode != SelectionMode.None && value is not null && (SelectedItem is null || !value.Equals(GetItemText(SelectedItem))))
```
Comment (AraHaan) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs:1134]: did you forget an ``is not`` here?

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs [hunk 9]
```diff
@@ -1976,7 +1976,7 @@ protected override void OnSelectedIndexChanged(EventArgs e)
             // calls CurrencyManager::EndCurrentEdit, and that will pull the dataFrom the controls
             // into the backEnd. We do not need to do that.
             //
-            if (DataManager != null && DataManager.Position != SelectedIndex)
+            if (DataManager is not null && DataManager.Position != SelectedIndex)
```
Comment (AraHaan) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs:1979]: ``is not SelectedIndex``?

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs [hunk 10]
```diff
@@ -1976,7 +1976,7 @@ protected override void OnSelectedIndexChanged(EventArgs e)
...
             {
                 //read this as "if everett or   (whidbey and selindex is valid)"
                 if (!FormattingEnabled || SelectedIndex != -1)
```
Comment (AraHaan) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs:1982]: ``is not -1``?

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs [hunk 11]
```diff
@@ -2019,7 +2019,7 @@ protected override void OnDisplayMemberChanged(EventArgs e)
             // we want to use the new DisplayMember even if there is no data source
             RefreshItems();
 
-            if (SelectionMode != SelectionMode.None && DataManager != null)
+            if (SelectionMode != SelectionMode.None && DataManager is not null)
```
Comment (AraHaan) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs:2022]: ``is not SelectionMode.None``

(There is possibly a lot more I missed on the if's)

**Review by AraHaan — CHANGES_REQUESTED** [review]

**Comment by RussKie** [comment]
> Cant you change all `!=` to `is not`, and all `==` to `is` where possible?

No. E.g. see https://stackoverflow.com/q/40676426/2338036
There are many more places in our codebase beyond `is null` and `is not null` where we could use `is` operator. However each of those "other" changes must be carefully reviewed.

**Review by RussKie — APPROVED** [review]

**Comment by RussKie** [comment]
Thank you

**Pull request merged**